### PR TITLE
Refresh unit cache command

### LINKED ISSUE #86: Refresh button for mech listings
I apologize if this is the incorrect place to post a feature request; I know this has been something excessively frustrating to myself and everyone who I avidly play with.  When you create or modify a BLK, the only option to update it in the mech listings is to reboot the game.  When you are making a lot of units (or trying funky things), this can get very frustrating - for example if I want to customize one unit in a company.

Long story short it would be HIGHLY useful to have a simple "Refresh" button that rescans the BLKs without having to completely close out and reload the game.

Again sorry if this is in the incorrect forum.  I just know this is something that over the years has probably caused considerable delays and general difficulty and it'd be great to see it updated.

### PR DESCRIPTION
**Pull request by neoancient**
This implements a request to be able to refresh the unit cache without restarting the program, using File>Unit List>Refresh unit cache. It works using the same process as loading the unit cache when MM starts, except that it starts with the current cache in memory instead of loading it from disk.

I also made a change in the way the name changes file is processed which was causing it to write a new units cache file ever when there were no changes. The way I wrote it was based on an assumption that the units cache was serialized as a map which included the lookup names. 

Closes #86

### CODE REVIEW

File: megamek/i18n/megamek/client/messages.properties [hunk 1]
```diff
@@ -663,6 +663,7 @@ CommonMenuBar.FileMenu=File
 CommonMenuBar.fileUnitsClear=Clear
 CommonMenuBar.fileUnitsReinforce=Reinforce from MUL
 CommonMenuBar.fileUnitsReinforceRAT=Reinforce from RAT
+CommonMenuBar.refreshUnitCache=Refresh unit cache
```
Comment (Windchild292) [inline megamek/i18n/megamek/client/messages.properties:666]: I'd prefer Refresh Unit Cache, as that follows the previous constants

File: megamek/src/megamek/client/ui/swing/CommonMenuBar.java [hunk 2]
```diff
@@ -313,6 +314,11 @@ public CommonMenuBar() {
         fileUnitsReinforceRAT.addActionListener(this);
         fileUnitsReinforceRAT.setActionCommand(ClientGUI.FILE_UNITS_REINFORCE_RAT);
         submenu.add(fileUnitsReinforceRAT);
+        fileRefreshCache = new JMenuItem(Messages
+                .getString("CommonMenuBar.refreshUnitCache")); //$NON-NLS-1$
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/CommonMenuBar.java:318]:
```suggestion
        fileRefreshCache = new JMenuItem(Messages.getString("CommonMenuBar.refreshUnitCache"));
```

File: megamek/src/megamek/client/ui/swing/CommonMenuBar.java [hunk 3]
```diff
@@ -313,6 +314,11 @@ public CommonMenuBar() {
         fileUnitsReinforceRAT.addActionListener(this);
         fileUnitsReinforceRAT.setActionCommand(ClientGUI.FILE_UNITS_REINFORCE_RAT);
         submenu.add(fileUnitsReinforceRAT);
+        fileRefreshCache = new JMenuItem(Messages
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/CommonMenuBar.java:317]: Can we please separate these? Might just be a personal preference, but I find it far easier to tell the differences between options when that's the case.

File: megamek/src/megamek/client/ui/swing/CommonMenuBar.java [hunk 4]
```diff
@@ -313,6 +314,11 @@ public CommonMenuBar() {
...
+                .getString("CommonMenuBar.refreshUnitCache")); //$NON-NLS-1$
+        fileRefreshCache.addActionListener(this);
+        fileRefreshCache.setActionCommand(ClientGUI.FILE_REFRESH_CACHE);
+        submenu.add(fileRefreshCache);
         fileUnitsOpen = new JMenuItem(Messages
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/CommonMenuBar.java:322]: Same here

File: megamek/src/megamek/client/ui/swing/CommonMenuBar.java (hunk 2 above)
Comment (Windchild292) [reply megamek/src/megamek/client/ui/swing/CommonMenuBar.java:318]: This falls under 100 with the removal of the comment

File: megamek/src/megamek/common/MechSummaryCache.java [hunk 5]
```diff
@@ -132,8 +152,8 @@ public void removeListener(Listener listener) {
     }
 
     private MechSummary[] m_data;
-    private Map<String, MechSummary> m_nameMap;
-    private Map<String, MechSummary> m_fileNameMap;
+    private final Map<String, MechSummary> m_nameMap;
+    private final Map<String, MechSummary> m_fileNameMap;
     private Map<String, String> hFailedFiles;
     private int cacheCount;
     private int fileCount;
```
Comment (Windchild292) [inline megamek/src/megamek/common/MechSummaryCache.java:159]: Pre-existing, but the location of these is odd. It should be at the top of the file.

File: megamek/src/megamek/common/MechSummaryCache.java [hunk 6]
```diff
@@ -236,6 +256,15 @@ public void loadMechData(boolean ignoreUnofficial) {
...
 
+        checkForChanges(ignoreUnofficial, vMechs, sKnownFiles, lLastCheck);
+        updateData(vMechs);
+        addLookupNames();
+        logReport();
+
+        done();
+    }
+
+    private void checkForChanges(boolean ignoreUnofficial, Vector<MechSummary> vMechs, Set<String> sKnownFiles, long lLastCheck) {
```
Comment (Windchild292) [inline megamek/src/megamek/common/MechSummaryCache.java:267]: Too long, let's split this into two lines

Comment (Windchild292) [reply megamek/src/megamek/common/MechSummaryCache.java:267]: Also, by the name I would assume it returns a boolean. Is there a reason this doesn't?

File: megamek/src/megamek/common/MechSummaryCache.java [hunk 7]
```diff
@@ -245,9 +274,23 @@ public void loadMechData(boolean ignoreUnofficial) {
...
 
+        // save updated cache back to disk
+        if (bNeedsUpdate) {
+            try {
+                saveCache();
+            } catch (Exception e) {
+                loadReport.append("  Unable to save mech cache\n");
+                MegaMek.getLogger().error(e);
+            }
+        }
```
Comment (Windchild292) [inline megamek/src/megamek/common/MechSummaryCache.java:285]: Possible location issue

File: megamek/src/megamek/common/MechSummaryCache.java [hunk 8]
```diff
@@ -245,9 +274,23 @@ public void loadMechData(boolean ignoreUnofficial) {
             bNeedsUpdate |= loadMechsFromDirectory(vMechs, sKnownFiles, lLastCheck, userDataUnits, ignoreUnofficial);
         }
 
+        // save updated cache back to disk
+        if (bNeedsUpdate) {
+            try {
+                saveCache();
+            } catch (Exception e) {
+                loadReport.append("  Unable to save mech cache\n");
```
Comment (Windchild292) [inline megamek/src/megamek/common/MechSummaryCache.java:282]:
```suggestion
                loadReport.append(" Unable to save mech cache\n");
```

Comment (Windchild292) [reply megamek/src/megamek/common/MechSummaryCache.java:282]: Double spacing is an accessibility issue. Further, I just want to check that the initial space is correct, as in general that's an odd way to handle it.

File: megamek/src/megamek/common/MechSummaryCache.java [hunk 9]
```diff
@@ -338,6 +361,38 @@ private void saveCache() throws Exception {
...
+                return;
+            }
+            File source = ms.getSourceFile();
+            if (source.exists()) {
+                units.add(ms);
+                if (null == ms.getEntryName()) {
+                    knownFiles.add(source.toString());
+                } else {
+                    knownFiles.add(ms.getEntryName());
+                }
```
Comment (Windchild292) [inline megamek/src/megamek/common/MechSummaryCache.java:383]: Ternary canditate

File: megamek/src/megamek/common/MechSummaryCache.java [hunk 10]
```diff
@@ -508,18 +563,18 @@ private boolean loadMechsFromDirectory(Vector<MechSummary> vMechs,
...
                 }
                 if (f.isDirectory()) {
-                    if (f.getName().toLowerCase().equals("unsupported")) {
+                    if (f.getName().equalsIgnoreCase("unsupported")) {
                         // Mechs in this directory are ignored because
                         // they have features not implemented in MM yet.
                         continue;
-                    } else if (f.getName().toLowerCase().equals("unofficial")
+                    } else if (f.getName().equalsIgnoreCase("unofficial")
                             && ignoreUnofficial) {
```
Comment (Windchild292) [inline megamek/src/megamek/common/MechSummaryCache.java:571]: Could be a single line

File: megamek/src/megamek/common/MechSummaryCache.java [hunk 11]
```diff
@@ -640,9 +695,8 @@ private boolean loadMechsFromZipFile(Vector<MechSummary> vMechs,
             ZipEntry zEntry = (ZipEntry) i.nextElement();
 
             if (zEntry.isDirectory()) {
-                if (zEntry.getName().toLowerCase().equals("unsupported")) {
-                    loadReport
-                            .append("  Do not place special 'unsupported' type folders in zip files, they must\n    be uncompressed directories to work properly.  Note that you may place\n    zip files inside of 'unsupported' type folders, though.\n");
+                if (zEntry.getName().equalsIgnoreCase("unsupported")) {
+                    loadReport.append("  Do not place special 'unsupported' type folders in zip files, they must\n    be uncompressed directories to work properly.  Note that you may place\n    zip files inside of 'unsupported' type folders, though.\n");
```
Comment (Windchild292) [inline megamek/src/megamek/common/MechSummaryCache.java:699]:
```suggestion
                    loadReport.append(" Do not place special 'unsupported' type folders in zip files, they must \nbe uncompressed directories to work properly. Note that you may place \nzip files inside of 'unsupported' type folders, though.\n");
```

**Review by Windchild292 — COMMENTED** [review]
In general this looks good, just a few things first

File: megamek/src/megamek/client/ui/swing/CommonMenuBar.java (hunk 3 above)
Comment (neoancient) [reply megamek/src/megamek/client/ui/swing/CommonMenuBar.java:317]: I'm not going to rework the formatting of this file for the sake of a small insertion.

File: megamek/src/megamek/common/MechSummaryCache.java (hunk 6 above)
Comment (neoancient) [reply megamek/src/megamek/common/MechSummaryCache.java]: I don't know why you would think that. There's no value to return.

File: megamek/src/megamek/common/MechSummaryCache.java (hunk 7 above)
Comment (neoancient) [reply megamek/src/megamek/common/MechSummaryCache.java]: I don't know what you mean by that.

File: megamek/src/megamek/common/MechSummaryCache.java (hunk 9 above)
Comment (neoancient) [reply megamek/src/megamek/common/MechSummaryCache.java:384]: Ternaries are good for assignments, but I find them clumsy as arguments.

File: megamek/src/megamek/common/MechSummaryCache.java (hunk 6 above)
Comment (Windchild292) [reply megamek/src/megamek/common/MechSummaryCache.java]: I would assume checkForChanges checks for any changes, but doesn't do any updates to it, just based on the name. Maybe add a comment saying it also updates, or rename?

File: megamek/src/megamek/client/ui/swing/CommonMenuBar.java (hunk 3 above)
Comment (Windchild292) [reply megamek/src/megamek/client/ui/swing/CommonMenuBar.java:317]: I meant just for this piece, not the whole file.

File: megamek/src/megamek/common/MechSummaryCache.java (hunk 7 above)
Comment (Windchild292) [reply megamek/src/megamek/common/MechSummaryCache.java]: As part of the previous comment on how I'd read the method name.

File: megamek/src/megamek/client/ui/swing/CommonMenuBar.java [hunk 12]
```diff
@@ -313,6 +314,11 @@ public CommonMenuBar() {
         fileUnitsReinforceRAT.addActionListener(this);
         fileUnitsReinforceRAT.setActionCommand(ClientGUI.FILE_UNITS_REINFORCE_RAT);
         submenu.add(fileUnitsReinforceRAT);
+        fileRefreshCache = new JMenuItem(Messages
+                .getString("CommonMenuBar.fileUnitsRefreshUnitCache")); //$NON-NLS-1$
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/CommonMenuBar.java:318]:
```suggestion
        fileRefreshCache = new JMenuItem(Messages.getString("CommonMenuBar.fileUnitsRefreshUnitCache"));
```

**Review by Windchild292 — APPROVED** [review]

**Pull request merged**